http/README.md

```diff
@@ -0,0 +1,133 @@
+  over the wire.
+
+- *HTTP transport* is the thing that takes your encoded reqeust/response and
+  sends it over the wire. For HTTP/1.1 and HTTP/2, this is a TCP+TLS connection.
+  For HTTP/3, this is a QUIC connection.
+
+When this document says *HTTP* it is generally referring to *HTTP semantics*.
+
+## Interoperability with existing HTTP systems
+
+A goal of this spec is to allow libp2p to be able to interoperate with existing HTTP servers and clients. Care is taken in this document to not introduce anything that would break interoperability with existing systems.
+
+## HTTP Transport
+
+Nodes MUST use HTTPS (i.e., they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.
+
+Nodes signal support for their HTTP transport using the `/http` component in
+their multiaddr. E.g., `/dns4/example.com/tls/http`. See the [HTTP multiaddr
+component spec](https://github.com/libp2p/specs/pull/550) for more details.
+
+## Namespace
+
+libp2p does not squat the global namespace. libp2p application protocols can be
+discovered by the [well-known resource](https://www.rfc-editor.org/rfc/rfc8615)
+`.well-known/libp2p`. This allows server operators to dynamically change the
+URLs of the application protocols offered, and not hard-code any assumptions how
+a certain resource is meant to be interpreted.
+
+```json
+
+{
+    "protocols": {
+        "/kad/1.0.0": {"path": "/kademlia/"},
+        "/ipfs/gateway": {"path": "/"},
+    }
+}
+```
+
+The resource contains a mapping of application protocols to a URL namespace. For
+example, this configuration file would tell a client
+
+1. That the Kademlia application protocol is available with prefix `/kademlia`
+and,
+2. The [IPFS Trustless Gateway API](https://specs.ipfs.tech/http-gateways/trustless-gateway/) is mounted at `/`.
+
+It is valid to expose a service at `/`. It is RECOMMENDED that implementations facilitate the coexistence of different service endpoints by ensuring that more specific URLs are resolved before less specific ones. For example, when registering handlers, more specific paths like `/kademlia/foo` should take precedence over less specific handler, such as `/`.
+
+## Peer ID Authentication
+
+When using the HTTP Transport, peer id authentication is optional. You only pay
+for it if you need it. This benefits use cases that don’t need peer
+authentication (e.g., fetching content addressed data) or authenticate some
+other way (not tied to libp2p peer ids).
+
+Specific authentications schemes for authenticating Peer IDs will be defined in
+a future spec.
+
+## Using HTTP semantics over stream transports
+
+Application protocols using HTTP semantics can run over any libp2p stream transport. Clients open a new stream using `/http/1.1` as the protocol identifer. Clients encode their HTTP request as an HTTP/1.1 message and send it over the stream. Clients parse the response as an HTTP/1.1 message and then close the stream.
```

Comment:
```suggestion
Application protocols using HTTP semantics can run over any libp2p stream transport. Clients open a new stream using `/http/1.1` as the protocol identifer. Clients encode their HTTP request as an HTTP/1.1 message and send it over the stream. Clients parse the response as an HTTP/1.1 message and then close the stream. Clients MUST NOT pipeline requests over a single stream.
```

Reply:
Why `MUST` ? There are valid use-cases for pipelining (when the responses only make sense in order).
It can be surprisingly hard to make existing HTTP implementation not reuse streams.

Reply:
This is only when using HTTP semantics on top of libp2p streams so existing implementations don't apply here. Right?

> There are valid use-cases for pipelining (when the responses only make sense in order).

Maybe, but it also complicates things. As a key datapoint HTTP/3 does not support request pipelining: "A client MUST send only a single request on a given stream." from https://datatracker.ietf.org/doc/html/rfc9114#name-expressing-http-semantics-i . This is consistent with that.

Reply:
> This is only when using HTTP semantics on top of libp2p streams so existing implementations don't apply here. Right?

I would expect nearly all libp2p-http implementations to use an existing http implementation like you did in `go-libp2p/p2p/http`.
I see your implementation does in fact not reuse stream however that because it's not compliant with `http1.1` you don't implement transfer encoding nor harden against miss-match of `Content-Length` and body (\*that what I found in 5 minutes of looking).
Your protocol does make sense as a simpler thing but then you can't call it `HTTP1.1` it's a subset of `HTTP1.1` where only closing the stream is used to indicate EOF (akin to HTTP1.0 but with `Host` headers).

The way I would have written `go-libp2p/p2p/http` is instead of implementing a custom HTTP transport, I would have used `http.Transport` with a custom `DialContext` handler and most of the features turned off (TLS, HTTP2, ...) aside of correctness, it also allow features like timeouts `req.Context`, `Accept-Encoding` for compression and others to work.

Reply:
> The way I would have written go-libp2p/p2p/http is instead of implementing a custom HTTP transport, I would have used http.Transport with a custom DialContext handler 

I think this would force you to serialize all HTTP requests, which is not a great default.

What's the use case that would benefit most from allowing pipelining? Can it not be addressed without pipelining?

Reply:
> I think this would force you to serialize all HTTP requests.

What do you mean ?

---

> What's the use case that would benefit most from allowing pipelining? Can it not be addressed without pipelining?

It helps resumptions of trustless gateway downloads.
You are in the middle of a file, you have a DFS stack of CIDs you need.
First I ask you for the inner most part of the dag I don't have, then the next one, then the next one.

If I do that without pipelining, I end up doing one round-trip for every unprocessed CID in the dag, which can easily be 50k+ if downloading files produced by optimized implementations.
Even with kubo defaults you can have to send thousands or so round trips in large ish files. (it gets worst as the number of links per block increase)

Instead of asking each CID one by one, I can use request pipelining to request for the next CIDs I want before I received the first ones.
The usual HTTP wisdom would be that this is bad because if a request stalls it blocks all the requests behind, but in this case I'm streaming the file I very precisely don't want the chunks out of order.

The alternatives are using multiple streams, which has worst resources control because the server will try to send me multiple blocks in parallel which I don't actually care about.
For this very particular usecase I could also recompute a single trustless gateway request which spans the undownloaded region of the files, however this is always gonna send me duplicate root blocks I already have, this is not future proof.

We don't gain anything server side, HTTP1.1 always makes it legal for the server to use EOF as indicating the end of a response (by not using `Content-Length` nor `Transfer-Encoding`).

Reply:
I'm not convinced by this use case.

First, it's important to point out that the **only** reason for pipelining in HTTP/1.1 is because you could not make a concurrent request without a whole new TCP connection. HTTP/2 and HTTP/3 dropped pipelining support because they introduced better ways to make concurrent requests.

This general problem of "I have a list of things I want to fetch in order" (which I believe is the same thing as your trustless gateway example) is not best solved through request pipelining. It is convenient to use a single channel and dump all your requests at the beginning, but there are downsides with this solution:

1. Request pipelining is not part of HTTP semantics. Therefore the application protocol should make no assumption on whether or not its request will be pipelined. The same application should run the same whether it's running on top of HTTP/1.1, h2, or h3. 
1. The request may fail at any point. This means there is more complicated logic besides "This request failed or suceeded." It also means you can't offload state to the request pipeline.
1. Sometimes the list of things you need changes (You seek somewhere else in the file), cancelling a single requests is a lot more straightforward than cancelling a whole stack of requests.
1. It's possible (and likely) that a set of concurrent request will be addressed faster than a set of serial requests, even if the client only cares about the responses in order. Imagine a 3/4 blocks are cached, but the last 1/4 needs to be fetched from a spinning disk.

The only thing that would convince me that we should support pipelining is if it would make implementations much simpler. My current experience with Go and JS is that that is not the case. And my brief survey into Rust leads me to think it's not necessary there either.

Not supporting pipelining for HTTP/1.1 encoding over libp2p streams does simplify implementations. One example: The library can safely close streams once it has sent a response. We avoid issues where the server closed a stream, but the client sent another request on that stream before it saw the close.

Reply:
I'm also going to add that the server SHOULD send back a `Connection: close` header. That should help enforce the idea that these connections over streams are not persistent.

Reply:
1. That looks like a me (client) problem, I don't see why the spec should disallow this.
2. I don't see why this is a problem, reset the stream and rematerialize the new request state from the unixfs DFS stack, I precisely don't want to track individual requests because a request might "fail" half way through but still send me useful blocks I was able to verify.
3. It's not, reset the whole stream and the whole stack is canceled. Then do rematerialization.
4. I agree it's faster, but it's less efficient, more buffer space is used, the server likely allocate more stacks, ... the RTT from the client to the server is often multiple orders worst than RTT from server to it's disks.
  This is way more tricky to implement efficiently, I can't double buffer it as if the blocks are small I'm not gonna fill up the pipe, so I need to track some concurrent requests window in my application layer. While with the pipelining solution I have a read and a write goroutine, and `.Write` is gonna block when the pipe is full, backpressure provided by libp2p.

> Not supporting pipelining for HTTP/1.1 encoding over libp2p streams does simplify implementations. One example: The library can safely close streams once it has sent a response. We avoid issues where the server closed a stream, but the client sent another request on that stream before it saw the close.

That not simplifying anything, the stream reusing is already an optional part of HTTP1.1, if you don't want to implement it send `Connection: close` and don't do it (https://github.com/libp2p/go-libp2p/pull/2733 aimed at proving this).

Reply:
If I change this from a MUST to a SHOULD would that alleviate your concerns here? I'm not convinced pipelining is a good thing here, but I can understand the desire that if a specific server and client want to do it, the spec shouldn't prohibit it. Server implementations can (SHOULD in the spec's term) send a `Connection: close` to not reuse streams.

Reply:
Yes thx